org-code-javabuilder/app/src/main/java/org/code/javabuilder/JavaRunner.java

```diff
@@ -50,40 +47,63 @@ public void compileAndRunUserProgram(UserProgram userProgram, Principal principa
     File tempFolder = null;
     try {
       tempFolder = Files.createTempDirectory("tmpdir").toFile();
+    } catch (IOException e) {
+      this.compileRunService.sendMessages(
+          principal.getName(),
+          "We hit an error on our side while compiling your program. Try again.");
+    }
 
-      this.compileRunService.sendMessages(principal.getName(), "Compiling your program...");
-      boolean compileSuccess = compileProgram(userProgram, principal, tempFolder);
-
-      if (compileSuccess) {
-        // this try statement will close the streams automatically
-        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
-            PrintStream out = new PrintStream(outputStream)) {
-          // set System.out to be a specific output stream in order to capture output of the
-          // program and send it back to the user
-          System.setOut(out);
-
-          this.compileRunService.sendMessages(principal.getName(), "Compiled!");
-          this.compileRunService.sendMessages(principal.getName(), "Running your program...");
-          runClass(tempFolder.toURI().toURL(), userProgram, principal);
-
-          // outputStream should now contain output of userProgram
-          outputStream.flush();
-          String result = outputStream.toString();
-          if (result.length() > 0) {
-            this.compileRunService.sendMessages(principal.getName(), result);
+    this.compileRunService.sendMessages(principal.getName(), "Compiling your program...");
+    boolean compileSuccess = compileProgram(userProgram, principal, tempFolder);
+
+    if (compileSuccess) {
+      this.compileRunService.sendMessages(principal.getName(), "Compiled!");
+      // We use a PipedInput and PipedOutput stream for I/O so we can control input to the user's
+      // program and output from the user's program.
+      // This try statement will close the streams automatically.
+      try (PipedInputStream inputStream = new PipedInputStream();
+          BufferedReader systemOutputReader =
+              new BufferedReader(new InputStreamReader(inputStream));
+          PipedOutputStream systemOutputStream = new PipedOutputStream(inputStream);
+          PrintStream out = new PrintStream(systemOutputStream)) {
+        // Set System.out to be a specific output stream in order to capture output of the
+        // program and send it back to the user
+        System.setOut(out);
+        this.systemInputStream = new PipedInputStream();
+        this.systemInputWriter = new PipedOutputStream(this.systemInputStream);
+        System.setIn(this.systemInputStream);
+
+        this.compileRunService.sendMessages(principal.getName(), "Running your program...");
+        JavaExecutorThread userRuntime =
+            new JavaExecutorThread(
+                tempFolder.toURI().toURL(), userProgram, principal, this.compileRunService);
+        userRuntime.start();
+
+        while (true) {
```

Comment:
could this condition be `userRuntime.isAlive()` or some variation on that? If not, can we put a comment above explaining why we have `while(true)` here? It would also be good to add a todo to kill the program if it runs for too long.

Reply:
Hah right. Fixed. Thanks! 
Here's the todo for infinite loops/long running programs: [CSA-151](https://codedotorg.atlassian.net/browse/CSA-151)